sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuArrowPythonRunner.scala

```diff
@@ -224,137 +147,102 @@ abstract class GpuArrowPythonRunnerBase(
         }
+        Table.writeArrowIPCChunked(builder.build(), new BufferToStreamWriter(dataOut))
+      }
 
-        Utils.tryWithSafeFinally {
-          while(inputIterator.hasNext) {
-            val table = withResource(inputIterator.next()) { nextBatch =>
-              GpuColumnVector.from(nextBatch)
-            }
-            withResource(new NvtxRange("write python batch", NvtxColor.DARK_GREEN)) { _ =>
-              // The callback will handle closing table and releasing the semaphore
-              writer.write(table)
-            }
+      var wrote = false
+      Utils.tryWithSafeFinally {
+        while (inputIterator.hasNext) {
+          wrote = false
+          val table = withResource(inputIterator.next()) { nextBatch =>
+            GpuColumnVector.from(nextBatch)
+          }
+          withResource(new NvtxRange("write python batch", NvtxColor.DARK_GREEN)) { _ =>
+            // The callback will handle closing table and releasing the semaphore
+            writer.write(table)
+            wrote = true
           }
-          // The iterator can grab the semaphore even on an empty batch
-          GpuSemaphore.releaseIfNecessary(TaskContext.get())
-        } {
-          writer.close()
-          dataOut.flush()
-          if (onDataWriteFinished != null) onDataWriteFinished()
         }
+        // The iterator can grab the semaphore even on an empty batch
+        GpuSemaphore.releaseIfNecessary(TaskContext.get())
+      } {
+        writer.close()
+        dataOut.flush()
+        if (onDataWriteFinished != null) onDataWriteFinished()
       }
+      wrote
+    }
 
-      private def writeEmptyIteratorOnCpu(dataOut: DataOutputStream): Unit = {
-        // most code is copied from Spark
-        val arrowSchema = ArrowUtilsShim.toArrowSchema(pythonInSchema, timeZoneId)
-        val allocator = ArrowUtils.rootAllocator.newChildAllocator(
-          s"stdout writer for empty partition", 0, Long.MaxValue)
-        val root = VectorSchemaRoot.create(arrowSchema, allocator)
-
-        Utils.tryWithSafeFinally {
-          val writer = new ArrowStreamWriter(root, null, dataOut)
-          writer.start()
-          // No data to write
-          writer.end()
-          // The iterator can grab the semaphore even on an empty batch
-          GpuSemaphore.releaseIfNecessary(TaskContext.get())
-        } {
-          root.close()
-          allocator.close()
-          if (onDataWriteFinished != null) onDataWriteFinished()
-        }
+    private def writeEmptyIteratorOnCpu(dataOut: DataOutputStream): Boolean = {
+      // most code is copied from Spark
+      val arrowSchema = ArrowUtilsShim.toArrowSchema(pythonInSchema, timeZoneId)
+      val allocator = ArrowUtils.rootAllocator.newChildAllocator(
+        s"stdout writer for empty partition", 0, Long.MaxValue)
+      val root = VectorSchemaRoot.create(arrowSchema, allocator)
+
+      val wrote = false
```

Comment:
Nit: This method could just return false directly as this is otherwise used.  It's also a bit weird to add a return type that is always hardcoded. Might be easier to follow if this didn't return a value as before, and instead the single caller of this returned false directly after calling this, since it's the caller of this that wants a value to return, not this method.